### PR TITLE
Zenparsing async generator functions

### PR DESCRIPTION
**Pull request by hzoo**
Ref https://github.com/babel/babel/pull/3473

Added an async exec test (had to modify the testing code to support async) from https://github.com/leebyron/async-to-gen

last commit: https://github.com/babel/babel/pull/4576/commits/b3bca86ef1f9019d563af4bb5c814faf3045cc54

exec: https://github.com/babel/babel/tree/b3bca86ef1f9019d563af4bb5c814faf3045cc54/packages/babel-plugin-transform-async-generator-functions/test/fixtures/for-await/async-generator-exec

cc @leebyron

### CODE REVIEW

File: packages/babel-core/src/transformation/file/index.js [hunk 1]
```diff
@@ -54,6 +54,7 @@ export default class File extends Store {
       sourceType:    this.opts.sourceType,
       filename:      this.opts.filename,
       sourceFileName:this.opts.filename,
+      allowReturnOutsideFunction: true,
```
Comment (hzoo) [inline packages/babel-core/src/transformation/file/index.js]: This is temporary until https://github.com/babel/babel/pull/3561 is ready and merged

File: packages/babel-helper-transform-fixture-test-runner/src/index.js [hunk 2]
```diff
@@ -151,7 +156,14 @@ export default function (
               return throwMsg === true || err.message.indexOf(throwMsg) >= 0;
             });
           } else {
-            runTask();
+            if (task.exec.code) {
+              let result = run(task);
+              if (result && typeof result.then === "function") {
+                return result;
```
Comment (hzoo) [inline packages/babel-helper-transform-fixture-test-runner/src/index.js]: Basically just returning the result of the `runExec(execOpts, execCode);` if it's a promise. We could also just always return

File: packages/babel-plugin-transform-async-generator-functions/test/fixtures/for-await/async-function/expected.js [hunk 3]
```diff
@@ -1,5 +1,5 @@
 let f = (() => {
-  var ref = babelHelpers.asyncToGenerator(function* () {
+  var _ref = babelHelpers.asyncToGenerator(function* () {
```
Comment (hzoo) [inline packages/babel-plugin-transform-async-generator-functions/test/fixtures/for-await/async-function/expected.js]: Not sure why this happened but some other code changed in the time between the original PR and now

File: packages/babel-plugin-transform-async-generator-functions/test/fixtures/for-await/async-generator-exec/exec.js [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+      fn(r);
+      return forEach(ai, fn);
+    }
+  });
+}
+
+var output = 0;
+return forEach(genAnswers(), function(val) { output += val.value })
+.then(function () {
+  assert.equal(output, 42);
```
Comment (hzoo) [inline packages/babel-plugin-transform-async-generator-functions/test/fixtures/for-await/async-generator-exec/exec.js]: `assert.equal(output, 42);` in a promise. Needs to use the `allowReturnOutsideFunction` parser option

File: packages/babel-core/src/transformation/file/index.js (hunk 1 above)
Comment (danez) [inline packages/babel-core/src/transformation/file/index.js]: The lets wait for #3561. I wouldn't add that now, as removing it again would be a breaking change if we are careful.

Comment (hzoo) [inline packages/babel-core/src/transformation/file/index.js]: Definitely - this was just to make it pass in this pr

File: packages/babel-helper-transform-fixture-test-runner/src/index.js (hunk 2 above)
Comment (loganfsmyth) [inline packages/babel-helper-transform-fixture-test-runner/src/index.js]: Do we need the `.then` check? I'd have thought we could return it without worrying to let Mocha worry about whether it is a promise.

File: packages/babel-helper-transform-fixture-test-runner/src/index.js [hunk 5]
```diff
@@ -151,7 +156,14 @@ export default function (
               return throwMsg === true || err.message.indexOf(throwMsg) >= 0;
             });
```
Comment (loganfsmyth) [inline packages/babel-helper-transform-fixture-test-runner/src/index.js]: Should we try to handle the promise case here too, to allow async tests to reject.

Comment (hzoo) [inline packages/babel-helper-transform-fixture-test-runner/src/index.js]: Do you want to do that in another pr if you want to and with an example test?
